contrib/drivers/netapp/ontap/ontap.go

```diff
@@ -0,0 +1,463 @@
+			"lunNumber":         publishInfo.IscsiLunNumber,
+			"igroup":            publishInfo.IscsiIgroup,
+		},
+	}
+
+	log.Infof("initialize connection success: %v", connInfo)
+	return connInfo, nil
+}
+
+func (d *Driver) TerminateConnection(opt *pb.DeleteVolumeAttachmentOpts) error {
+	var name = getVolumeName(opt.GetVolumeId())
+
+	// Validate Flexvol exists before trying to Unmount
+	volExists, err := d.sanStorageDriver.API.VolumeExists(name)
+	if err != nil {
+		return fmt.Errorf("error checking for existing volume: %v", err)
+	}
+	if !volExists {
+		log.Infof("Volume %s already deleted, skipping destroy.", name)
+		return nil
+	}
+
+	// Unmount the FlexVolume
+	volUnmountResponse, err := d.sanStorageDriver.API.VolumeUnmount(name, true)
+	if err != nil {
+		return fmt.Errorf("error destroying volume %v: %v", name, err)
+	}
+	if zerr := api.NewZapiError(volUnmountResponse); !zerr.IsPassed() {
+		return fmt.Errorf("error destroying volume %v: %v", name, zerr.Error())
+	}
+
+	log.Infof("termination connection success")
+	return nil
+}
+
+func (d *Driver) CreateSnapshot(opt *pb.CreateVolumeSnapshotOpts) (snap *model.VolumeSnapshotSpec, err error) {
+	var snapName = getSnapshotName(opt.GetId())
+	var volName = getVolumeName(opt.GetVolumeId())
+
+	snapConfig := d.GetSnapshotConfig(snapName, volName)
+
+	snapshot, err := d.sanStorageDriver.CreateSnapshot(snapConfig)
+
+	if err != nil {
+		msg := fmt.Sprintf("create snapshot %s (%s) failed: %s", opt.GetName(), opt.GetId(), err)
+		log.Error(msg)
+		return nil, err
+	}
+
+	log.Infof("Create Snapshot %s (%s) Success.", opt.GetName(), opt.GetId())
+
+	return &model.VolumeSnapshotSpec{
+		BaseModel: &model.BaseModel{
+			Id: opt.GetId(),
+		},
+		Name:        opt.GetName(),
+		Description: opt.GetDescription(),
+		VolumeId:    opt.GetVolumeId(),
+		Size:        opt.GetSize(),
+		Metadata: map[string]string{
+			"name":         snapName,
+			"volume":       volName,
+			"creationTime": snapshot.Created,
+			"size":         strconv.FormatInt(snapshot.SizeBytes/bytesGB, 10) + "GB",
+		},
+	}, nil
+}
+
+func (d *Driver) PullSnapshot(snapIdentifier string) (*model.VolumeSnapshotSpec, error) {
+	// not used, do nothing
+	return nil, &model.NotImplementError{"method PullSnapshot has not been implemented yet"}
+}
+
+func (d *Driver) DeleteSnapshot(opt *pb.DeleteVolumeSnapshotOpts) error {
+
+	var snapName = getSnapshotName(opt.GetId())
+	var volName = getVolumeName(opt.GetVolumeId())
+
+	snapConfig := d.GetSnapshotConfig(snapName, volName)
+
+	err := d.sanStorageDriver.DeleteSnapshot(snapConfig)
+
+	if err != nil {
+		msg := fmt.Sprintf("delete volume snapshot (%s) failed: %v", opt.GetId(), err)
+		log.Error(msg)
+		return err
+	}
+	log.Infof("remove volume snapshot (%s) success", opt.GetId())
+	return nil
+}
+
+func (d *Driver) ListPools() ([]*model.StoragePoolSpec, error) {
+
+	var pools []*model.StoragePoolSpec
+
+	aggregates, err := d.sanStorageDriver.API.VserverGetAggregateNames()
```

Comment:
This seems to be generic to the sds requiring vserver. Here the file is ontap.go which appears to be generic.